File: scripts/src/contracts/tacoclicker/schemas.ts
import { schemaAlkaneId } from "tacoclicker-sdk";
import { BorshSchema, Infer as BorshInfer } from "borsher";

export const schemaTacoClickerConsts = BorshSchema.Struct({
  controlled_mint_factory: schemaAlkaneId,
  tortilla_alkane_id: schemaAlkaneId,
});

export type IAlkaneId = BorshInfer<typeof schemaAlkaneId>;

export const schemaTaqueriaParam = BorshSchema.Struct({
  taqueria: schemaAlkaneId,
});
export type ITaqueriaParam = BorshInfer<typeof schemaTaqueriaParam>;

/*────────────────────────────  Request / Response Schemas  ────────────────────*/

// 105  GetConsts → Vec<u8>  (opaque bytes)  – no additional TS schema needed.

// 106  Register            → Vec<u8>        – same, opaque.

// 107  GetTaqueriaFromAlkaneList
export const schemaAlkaneList = BorshSchema.Struct({
  alkanes: BorshSchema.Vec(schemaAlkaneId),
});
export type IAlkaneList = BorshInfer<typeof schemaAlkaneList>;

// 108  GetTortillaId       → SchemaAlkaneId  (reuse schemaAlkaneId)

// 110  GetTortillaPerBlockForTaqueria
export const schemaTortillaPerBlockRes = BorshSchema.Struct({
  tortilla_per_block: BorshSchema.u128,
});
export type ITortillaPerBlockRes = BorshInfer<typeof schemaTortillaPerBlockRes>;

// 111  GetUnclaimedTortillaForTaqueria
export const schemaUnclaimedRes = BorshSchema.Struct({
  unclaimed_tortilla: BorshSchema.u128,
});
export type IUnclaimedRes = BorshInfer<typeof schemaUnclaimedRes>;

// 112  GetUpgradesForTaqueria   → SchemaUserUpgradesView (big struct)
export const schemaUserUpgradesEntry = BorshSchema.Struct({
  amount: BorshSchema.u128,
  next_price: BorshSchema.u128,
});
export const schemaUserUpgradesView = BorshSchema.Struct({
  taquero: schemaUserUpgradesEntry,
  salsa_bar: schemaUserUpgradesEntry,
  tortilla_tree: schemaUserUpgradesEntry,
  tortilla_factory: schemaUserUpgradesEntry,
  taco_bank: schemaUserUpgradesEntry,
  taco_pyramid: schemaUserUpgradesEntry,
  tortilla_spaceship: schemaUserUpgradesEntry,
  satoshi_tacomoto: schemaUserUpgradesEntry,
});
export type IUserUpgradesView = BorshInfer<typeof schemaUserUpgradesView>;

// 113  GetAvailableUpgrades    → SchemaUpgradesView
export const schemaUpgradesEntry = BorshSchema.Struct({
  cost: BorshSchema.u128,
  weight: BorshSchema.u128,
  current_emission: BorshSchema.u128,
});
export const schemaUpgradesView = BorshSchema.Struct({
  taquero: schemaUpgradesEntry,
  salsa_bar: schemaUpgradesEntry,
  tortilla_tree: schemaUpgradesEntry,
  tortilla_factory: schemaUpgradesEntry,
  taco_bank: schemaUpgradesEntry,
  taco_pyramid: schemaUpgradesEntry,
  tortilla_spaceship: schemaUpgradesEntry,
  satoshi_tacomoto: schemaUpgradesEntry,
});
export type IUpgradesView = BorshInfer<typeof schemaUpgradesView>;

// 114  GetMultiplierFromHash
export const schemaGetMulReq = BorshSchema.Struct({
  hash_bytes_be: BorshSchema.Vec(BorshSchema.u8),
});
export const schemaGetMulRes = BorshSchema.Struct({
  multiplier: BorshSchema.u128,
});
export type IGetMulRes = BorshInfer<typeof schemaGetMulRes>;

// 115  BuyUpgrade
export const schemaBuyUpgradeReq = BorshSchema.Struct({
  upgrade: BorshSchema.u8, // UpgradeKind enum discriminant
});
// returns nothing (just a success flag in the CallResponse)

// 116  BetOnBlock
export const schemaBetOnBlockReq = BorshSchema.Struct({
  nonce_found_poc: BorshSchema.u128,
  target_multiplier: BorshSchema.u128,
});
export const schemaBetOnBlockRes = BorshSchema.Struct({
  won_amount: BorshSchema.u128,
  lost_amount: BorshSchema.u128,
});

File: contracts/tacoclicker/src/game/schemas.rs
use borsh::{BorshDeserialize, BorshSerialize};

use crate::schemas::SchemaAlkaneId;

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaGlobalEmissionState {
    pub total_weight: u128,
    pub acc_reward_per_weight: u128,
    pub last_updated_block: u128,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaTaqueriaEmissionState {
    pub taqueria_weight: u128,
    pub reward_debt: u128,
    pub pending: u128,
    pub last_poc_hash: Vec<u8>,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaUpgradesEntry {
    pub cost: u128,
    pub weight: u128,
    pub current_emission: u128,
}
#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaUpgradesView {
    pub taquero: SchemaUpgradesEntry,
    pub salsa_bar: SchemaUpgradesEntry,
    pub tortilla_tree: SchemaUpgradesEntry,
    pub tortilla_factory: SchemaUpgradesEntry,
    pub taco_submarine: SchemaUpgradesEntry,
    pub taco_pyramid: SchemaUpgradesEntry,
    pub tortilla_spaceship: SchemaUpgradesEntry,
    pub satoshi_tacomoto: SchemaUpgradesEntry,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaUserUpgradesEntry {
    pub amount: u128,
    pub next_price: u128, //Increases by 1.5x on each buy. 50000, 75000, etc etc. This is to incentivize people to chase bigger upgrades
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaGlobalSalsaState {
    pub current_block: u128,
    pub best_hash: Vec<u8>,
    pub best_hash_owner: Vec<u8>, //Increases by 1.5x on each buy. 50000, 75000, etc etc. This is to incentivize people to chase bigger upgrades
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaUserUpgradesView {
    pub taquero: SchemaUserUpgradesEntry,
    pub salsa_bar: SchemaUserUpgradesEntry,
    pub tortilla_tree: SchemaUserUpgradesEntry,
    pub tortilla_factory: SchemaUserUpgradesEntry,
    pub taco_submarine: SchemaUserUpgradesEntry,
    pub taco_pyramid: SchemaUserUpgradesEntry,
    pub tortilla_spaceship: SchemaUserUpgradesEntry,
    pub satoshi_tacomoto: SchemaUserUpgradesEntry,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug, Clone, Copy)]
pub enum UpgradeKind {
    Taquero,
    SalsaBar,
    TortillaTree,
    TortillaFactory,
    TacoBank,
    TacoPyramid,
    TortillaSpaceship,
    SatoshiTacomoto,
}

impl From<UpgradeKind> for u8 {
    fn from(kind: UpgradeKind) -> Self {
        match kind {
            UpgradeKind::Taquero => 0,
            UpgradeKind::SalsaBar => 1,
            UpgradeKind::TortillaTree => 2,
            UpgradeKind::TortillaFactory => 3,
            UpgradeKind::TacoBank => 4,
            UpgradeKind::TacoPyramid => 5,
            UpgradeKind::TortillaSpaceship => 6,
            UpgradeKind::SatoshiTacomoto => 7,
        }
    }
}

//buy_upgrade
#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaBuyUpgradeParameters {
    pub upgrade: UpgradeKind,
}

//get available upgrades
#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]

pub struct SchemaTaqueriaSpecificParameters {
    pub taqueria: SchemaAlkaneId,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]

pub struct SchemaGetMultiplierFromHashParameters {
    pub hash_bytes_be: Vec<u8>,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]

pub struct SchemaGetMultiplierFromHashResponse {
    pub multiplier: u128,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]

pub struct SchemaGetUnclaimedTortillaResponse {
    pub unclaimed_tortilla: u128,
}
#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]

pub struct SchemaGetTortillaPerBlockResponse {
    pub tortilla_per_block: u128,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]

pub struct SchemaBetOnBlockParameters {
    pub nonce_found_poc: u128,
    pub target_multiplier: u128,
}

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]

pub struct SchemaBetOnBlockResponse {
    pub won_amount: u128,
    pub lost_amount: u128,
}
#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
pub struct SchemaCompleteGlobalState {
    pub emission_state: SchemaGlobalEmissionState,
    pub salsa_state: SchemaGlobalSalsaState,
}
